### PR TITLE
MSRIOGroup refactor

### LINKED ISSUE #435: Refactor PlatformIO to move POWER and TEMPERATURE into MSRIOGroup
## Done Criteria:
- Ensure that these signals are provided by the MSRIOGroup - TEMPERATURE_CORE, TEMPERATURE_PKG, POWER_PKG, POWER_DRAM
- Add relevant unit tests
- Ensure that the error messaging related to signal-dependencies is correct

## Related Bugs:
- #997
- #821 

This also allows for #377 to be implemented by MSRIOGroup only instead of using PlatformIO to combine MSRs.

### LINKED ISSUE #821: geopmread unable to find domain of available controls/signals
```
$ ./geopmread
CPUINFO::FREQ_MAX
CPUINFO::FREQ_MIN
CPUINFO::FREQ_STEP
CPUINFO::FREQ_STICKER
FREQUENCY_MIN
FREQUENCY_STEP
FREQUENCY_STICKER
POWER_DRAM
POWER_PACKAGE
TEMPERATURE_CORE
TEMPERATURE_PACKAGE
TIME
TIME::ELAPSED
$ ./geopmread -d POWER_PACKAGE
Error: unable to determine signal type: <geopm> Invalid argument: PlatformIOImp::signal_domain_type(): signal name "ENERGY_PACKAGE" not found: at geopm/../src/PlatformIO.cpp:190
$ ./geopmread -d TEMPERATURE_CORE
Error: unable to determine signal type: <geopm> Invalid argument: PlatformIOImp::signal_domain_type(): signal name "TEMPERATURE_CORE_UNDER" not found: at geopm/../src/PlatformIO.cpp:190
```
The non-power, non-temperature entries behave as expected.  Running on an (unsupported?) 0x9E i9-9900K processor.

### LINKED ISSUE #997: Confusing error message for invalid signal name with POWER_ prefix
The signal "POWER_PACKAGE_LIMIT" does not exist, but the error message indicates that a different (non-existent) MSR is the problem.
```
> geopmread POWER_PACKAGE_LIMIT package 0
Error: cannot read signal: <geopm> Invalid argument: PlatformIOImp::read_signal(): signal name "ENERGY_PACKAGE_LIMIT" not found: at geopm/src/PlatformIO.cpp:545
```
We probably need more strict name checking for the custom power and temperature signals that live in PlatformIO, or fix it for good with #435

### LINKED ISSUE #1040: Change MSRIOGroup to use Signal interface for single MSR-based signal
> estimate 5

Change MSRIOgroup to use vectors of Signal type rather than MSR/MSRSignal vectors.
Use the new related interfaces in MSRIO to let MSRIO own the memory for signal values; remove raw pointers from MSRIOGroup and Signal implementations.

Done criteria
--------------
- Signals for raw MSRs, MSR-field-based signals, and temperature are provided by MSRIOGroup through Signal objects.
- Unit test covering all MSR's that appear in the trace and report

Parent story #435

### LINKED ISSUE #1042: Add power signals to MSRIOGroup
> estimate 2

Change MSRIOgroup to use DerviativeSignal to support the power signals from PlatformIO. 

Done criteria
--------------
- POWER_PACKAGE POWER_DRAM are supported by MSRIOGroup directly
- add unit tests for power signals in the MSRIOGroup unit tests.

Parent story #435

### PR DESCRIPTION
**Pull request by dianarg**
- Add more unit tests.
- Use Signal classes to implement MSRIOGroup
- New methods in MSRIO interface to support refactoring.
- Remove old implementation from PlatformIO.
- Closes #435 
- Closes #1040 
- Closes #1042 
- Fixes #997 
- Fixes #821

### CODE REVIEW

File: src/MSRIOGroup.cpp [hunk 1]
```diff
@@ -151,63 +151,131 @@ namespace geopm
...
+                    auto eng = readings[domain_idx];
+                    result[domain_idx] =
+                        std::make_shared<DerivativeSignal>(time_sig, eng,
+                                                           num_read_loops, sleep_time);
+                }
+                m_signal_available[signal_name] = result;
+                m_signal_domain[signal_name] = energy_domain;
+                m_signal_units[signal_name] = MSR::M_UNITS_WATTS;
+                m_signal_agg_func[signal_name] = agg_function(msr_name);
+                // @todo: format
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: these todos can be removed; handled elsewhere

File: test/geopm_test_helper.cpp [hunk 2]
```diff
@@ -69,3 +70,36 @@ bool is_format_raw64(std::function<std::string(double)> func)
     std::string expected = "0x003ff00000000000";
     return func(value) == expected;
 }
+
+static std::vector<double> example_data = {1, 2, 4, 10};
+
+bool is_agg_sum(std::function<double(const std::vector<double> &)> func)
+{
+    return func(example_data) == geopm::Agg::sum(example_data);
```
Comment (cmcantalupo) [inline test/geopm_test_helper.cpp:78]: could be 

   return (func.target == geopm::Agg::sum);

But this seems fine as it

File: test/MSRIOGroupTest.cpp [hunk 3]
```diff
@@ -222,55 +235,187 @@ TEST_F(MSRIOGroupTest, supported_cpuid)
...
+    // counter
+    func = m_msrio_group->agg_function("INSTRUCTIONS_RETIRED");
+    EXPECT_TRUE(is_agg_sum(func));
+    func = m_msrio_group->agg_function("CYCLES_THREAD");
+    EXPECT_TRUE(is_agg_sum(func));
+    func = m_msrio_group->agg_function("CYCLES_REFERENCE");
+    EXPECT_TRUE(is_agg_sum(func));
+    /// @todo: what should this be?
+    //func = m_msrio_group->agg_function("TIMESTAMP_COUNTER");
+    //EXPECT_TRUE(is_agg_sum(func));
```
Comment (dianarg) [inline test/MSRIOGroupTest.cpp:352]: see #1058

File: test/MSRIOGroupTest.cpp [hunk 4]
```diff
@@ -222,55 +235,187 @@ TEST_F(MSRIOGroupTest, supported_cpuid)
...
+    std::vector<std::string> si_alias = {
+        "ENERGY_PACKAGE", "ENERGY_DRAM",
+        "FREQUENCY", "FREQUENCY_MAX",
+        "MSR::TEMPERATURE_CORE", "MSR::TEMPERATURE_PACKAGE",
+        "POWER_PACKAGE_MIN", "POWER_PACKAGE_MAX", "POWER_PACKAGE_TDP"
+    };
+    for (const auto &name : si_alias) {
+        func = m_msrio_group->format_function(name);
+        EXPECT_TRUE(is_format_double(func));
+
```
Comment (dianarg) [inline test/MSRIOGroupTest.cpp]: extra line

File: src/MSRIOGroup.cpp [hunk 5]
```diff
@@ -168,46 +171,116 @@ namespace geopm
         register_msr_control("POWER_PACKAGE_LIMIT", "MSR::PKG_POWER_LIMIT:PL1_POWER_LIMIT");
         register_msr_control("FREQUENCY",        "MSR::PERF_CTL:FREQ");
         register_msr_control("POWER_PACKAGE_TIME_WINDOW", "MSR::PKG_POWER_LIMIT:PL1_TIME_WINDOW");
+
+        // @todo fix order of member construction
     }
 
-    void MSRIOGroup::register_raw_msr_signal(const std::string &msr_name, const MSR &msr_ptr)
+    MSRIOGroup::~MSRIOGroup()
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: could use = default

File: src/MSRIOGroup.cpp [hunk 6]
```diff
@@ -168,46 +171,116 @@ namespace geopm
...
+                auto readings = read_it->second;
+                int read_domain = m_signal_domain[msr_name];
+                int num_domain = m_platform_topo.num_domain(read_domain);
+                GEOPM_DEBUG_ASSERT(num_domain == (int)readings.size(),
+                                   "size of domain for " + msr_name +
+                                   " does not match number of signals available.");
+                std::vector<std::shared_ptr<Signal> > result(num_domain);
+                for (int domain_idx = 0; domain_idx < num_domain; ++domain_idx) {
+                    int max_idx = *m_platform_topo.domain_nested(max_domain, read_domain, domain_idx).begin();
+                    auto min = max_it->second[max_idx];
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: confusing name

File: src/MSRIOGroup.cpp [hunk 7]
```diff
@@ -168,46 +171,116 @@ namespace geopm
...
+            {"MSR::TEMPERATURE_CORE", "MSR::THERM_STATUS:DIGITAL_READOUT"},
+            {"MSR::TEMPERATURE_PACKAGE", "MSR::PACKAGE_THERM_STATUS:DIGITAL_READOUT"}
+        };
+        for (const auto &ts : temp_signals) {
+            std::string signal_name = ts.first;
+            std::string msr_name = ts.second;
+            auto read_it = m_signal_available.find(msr_name);
+            if (max_it != m_signal_available.end() &&
+                read_it != m_signal_available.end())
+            {
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: move up

File: src/MSRIOGroup.cpp [hunk 8]
```diff
@@ -168,46 +171,116 @@ namespace geopm
...
+                    auto dim = readings[domain_idx];
+                    result[domain_idx] = std::make_shared<DifferenceSignal>(min, dim);
+                }
+                m_signal_available[signal_name] = result;
+                m_signal_domain[signal_name] = read_domain;
+                m_signal_units[signal_name] = MSR::M_UNITS_CELSIUS;
+                m_signal_agg_func[signal_name] = agg_function(msr_name);
+            }
+        }
+        m_signal_desc_map["MSR::TEMPERATURE_CORE"] = "Core temperature in degrees C";
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: could move up into loop if temp_signals also contained description

File: src/MSRIOGroup.cpp [hunk 9]
```diff
@@ -168,46 +171,116 @@ namespace geopm
...
+        // register time signal; domain board
+        std::string time_name = "MSR::TIME";
+        std::shared_ptr<Signal> time_sig = std::make_shared<TimeSignal>(m_time_zero, m_time_batch);
+        m_signal_available[time_name] = std::vector<std::shared_ptr<Signal> >({time_sig});
+        m_signal_domain[time_name] = GEOPM_DOMAIN_BOARD;
+        m_signal_units[time_name] = MSR::M_UNITS_SECONDS;
+        m_signal_desc_map[time_name] = "Time in seconds used to calculate power";
+
+        // hardcoded parameters for read_signal for power
+        int num_read_loops = 8;
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: derivative_window

File: src/MSRIOGroup.cpp [hunk 10]
```diff
@@ -168,46 +171,116 @@ namespace geopm
...
+        double sleep_time = 0.005;  // 5000 us
+
+        // Power signal name with corresponding energy signal.  The
+        // domain will match that of the energy signal.
+        std::vector<std::pair<std::string, std::string> > power_signals {
+            {"MSR::POWER_PACKAGE", "ENERGY_PACKAGE"},
+            {"MSR::POWER_DRAM", "ENERGY_DRAM"}
+        };
+        m_signal_desc_map["MSR::POWER_PACKAGE"] = "Average package power in watts over the last 80 ms";
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: 40 ms not accurate for push_signal for some agents.
"Average package power in watts over a window of 40ms or 8 control loop iterations"

File: src/MSRIOGroup.cpp [hunk 11]
```diff
@@ -752,66 +708,32 @@ namespace geopm
         m_is_active = true;
     }
 
-    void MSRIOGroup::register_msr_signal(const std::string &msr_name)
-    {
-        register_msr_signal(msr_name, msr_name);
-    }
-
     void MSRIOGroup::register_msr_signal(const std::string &signal_name, const std::string &msr_name_field)
```
Comment (dianarg) [inline src/MSRIOGroup.cpp]: name that reflects this is for aliases

File: src/MSRIOGroup.hpp [hunk 12]
```diff
@@ -127,13 +121,45 @@ namespace geopm
...
+            void register_temperature_signals(void);
+            /// @brief Add support for power combined signals if underlying
+            ///        signals are available.
+            void register_power_signals(void);
+
+            // all available signals: map from name and domain index
+            // to Signal ptr.  This pointer should be copied when signal
+            // is pushed and used directly for read_signal.
+            std::map<std::string, std::vector<std::shared_ptr<Signal> > > m_signal_available;
+            std::map<std::string, int> m_signal_domain;
```
Comment (dianarg) [inline src/MSRIOGroup.hpp]: could be a struct

File: src/MSRIOGroup.hpp [hunk 13]
```diff
@@ -144,22 +170,15 @@ namespace geopm
...
             std::vector<bool> m_is_adjusted;
-            // TODO: figure out diff with m_name_msr_map
             std::vector<std::unique_ptr<MSR> > m_msr_arr;
-            // Mappings from names to all valid signals and controls
+            // Mappings from names to all valid controls
             std::map<std::string, const MSR &> m_name_msr_map;
-            std::map<std::string, std::vector<std::shared_ptr<MSRSignal> > > m_name_cpu_signal_map;
             std::map<std::string, std::vector<std::shared_ptr<MSRControl> > > m_name_cpu_control_map;
-            // Pushed signals and controls only
-            std::vector<std::shared_ptr<MSRSignal> > m_active_signal;
```
Comment (dianarg) [inline src/MSRIOGroup.hpp:157]: okay to get rid of MSRSignal class
replaced by RawMSRSignal, MSRFieldSignal

**Review by dianarg — COMMENTED** [review]
- Have register_alias skip the alias instead of throwing if underlying MSR signal can't be found
- Can remove DifferenceConbinedSignal and DerivativeCombinedSignal

**Review by cmcantalupo — APPROVED** [review]
LGTM

**Pull request merged**